File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.java
class Solution {
    public boolean allzero(int [] freq){
        for(int ele : freq){
            if(ele!=0) return false;
        }
        return true;
    }
    public List<Integer> findAnagrams(String s, String p) {
        int k =p.length();
        int n = s.length();
         List<Integer>ans = new ArrayList<>();
        if(n<k) return ans;
        int[] freq = new int[26];
        for(int i = 0;i<k;i++){
            freq[p.charAt(i)-'a']++;
        }
       
        //checking first window
        for(int i = 0;i<k;i++){
            freq[s.charAt(i)-'a']--;
        }
        if(allzero(freq)){
            ans.add(0);
        }

        //now chacking remaining windows
        for(int i =k;i<n;i++){
            freq[s.charAt(i)-'a']--; //--means we are addind
            freq[s.charAt(i-k)-'a']++; //++ means we are removind
            if(allzero(freq)){
                ans.add(i-k+1);
            }
        }
        return ans;

    }
}

File: 0438-find-all-anagrams-in-a-string/0438-find-all-anagrams-in-a-string.cpp
class Solution {
public:
    bool allzero(vector<int>v){
        for(int ele : v){
            if(ele!=0) return false;
        }
        return true;
    }
    vector<int> findAnagrams(string s, string p) {
        int k = p.size();
        int n = s.size();
        if (n < k) return {}; // Edge case: s is shorter than p
        vector<int> v (26,0);
        for(int i=0;i<k;i++){
            v[p[i]-'a']++;
        }
        
        vector<int>ans;

        for (int i = 0; i < k; i++) { //first window
            v[s[i] - 'a']--;
        }
        // Check the first window
        if (allzero(v)) {
            ans.push_back(0);
        }
        //remainig windows
        
        for (int i = k; i < n; i++) {
            v[s[i] - 'a']--;        // Add new character to window
            v[s[i - k] - 'a']++;    // Remove old character from window

            if (allzero(v)) {
                ans.push_back(i - k + 1);
            }
        }
        return ans;
    }
};